### PR TITLE
How to get SSL/HTTPS for localhost

### PR DESCRIPTION
**Pull request by lewe01**
How to get SSL/HTTPS for localhost

### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

File: articles/how-to-get-ssl-https-for-localhost/index.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+$ npm i express https
+$ touch index.js
+```
+
+Then edit the `index.js` file as below:
+
+```JavaScript
+const fs = require('fs');
+const key = fs.readFileSync('../localhost/localhost.decrypted.key');
+const cert = fs.readFileSync('../localhost/localhost.crt');
```
Comment (adrianmurage) [inline articles/how-to-get-ssl-https-for-localhost/index.md:123]: Point out that the path for the `localhost.decrypted.key` and `localhost.crt` are dependent on where the reader is running the command from.

For me it was 
```JavaScript
const key = fs.readFileSync('./CA/localhost/localhost.decrypted.key');
const cert = fs.readFileSync('./CA/localhost/localhost.crt');
```

File: articles/how-to-get-ssl-https-for-localhost/index.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+![postman red security error](/engineering-education/hot-to-get-ssl-https-for-localhost/postman-red-security-error.PNG)
+
+To solve this issue, we can add our CA to Postman. In Postman settings, make sure that CA certificates are turned on and select the `CA.pem` file.
+Then rerun the request as below; this time, we will have a secure **200 OK** message.
+
+![postman success cert](/engineering-education/hot-to-get-ssl-https-for-localhost/postman-success-cert.PNG)
+
+### Conclusion
+In this guide, we have seen it is possible to have localhost served up with SSL. From the guide, we have been able to set the CA and used it to sign the SSL certificate. We changed the browser settings as well as the API client settings for them to accept the certificate. Note it is never advisable to add certificates to the certificate list thoughtlessly. Malicious users everywhere will pose as a CA and trick the user into
+accepting their websites posing as authentic, leading to a serious security breach. The process can help developers host their sites on local servers or forwarded ports to assure their visitors that the site is safe and secure.
```
Comment (adrianmurage) [inline articles/how-to-get-ssl-https-for-localhost/index.md:197]:
```suggestion
In this guide, we have seen it is possible to have localhost served up with SSL. 

From the guide, we have been able to set the CA and used it to sign the SSL certificate. We changed the browser settings as well as the API client settings for them to accept the certificate.

Note it is never advisable to add certificates to the certificate list thoughtlessly. Malicious users everywhere will pose as a CA and trick the user into accepting their websites posing as authentic, leading to a serious security breach.

The process can help developers host their sites on local servers or forwarded ports to assure their visitors that the site is safe and secure.
```

File: articles/how-to-get-ssl-https-for-localhost/index.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
+### Introduction
+Building a web application that later can be deployed in the production server requires first to be tested in a local environment. It is also possible to run the web application in a local environment and be accessed over the internet. It is equally important that we get an [SSL certificate](https://www.cloudflare.com/en-gb/learning/ssl/what-is-an-ssl-certificate/) for the same, for the users to trust the site and remove the red annoying *'Not Secure'* message at the address bar.
```
Comment (adrianmurage) [inline articles/how-to-get-ssl-https-for-localhost/index.md:2]:
```suggestion
Building a web application for a production server requires local testing before deployment. 

It is also possible to run the web application in a local environment have it accessible over the internet.

It is equally important that we get an [SSL certificate](https://www.cloudflare.com/en-gb/learning/ssl/what-is-an-ssl-certificate/) for the same, for the users to trust the site and to remove the *'Not Secure'* message at the address bar.

This article will take you through how to provision SSL certificates for local servers.
```

**Review by adrianmurage — CHANGES_REQUESTED** [review]
Welcome to EngEd @lewe01 
This is a great first contribution!

In my review, I pointed out a few changes. Go through them and make the necessary changes.
@ me when you're done

**Comment by lewe01** [comment]
@adrianmurage The requested changes has been done.

**Comment by adrianmurage** [comment]
@lewe01 Great job with the article!
Looking forward to seeing more contributions

**Comment by lewe01** [comment]
@adrianmurage Thank you

**Comment by adrianmurage** [comment]
@hectorkambow this article is ready for final review

**Pull request merged**